### PR TITLE
[Form][FrameworkBundle][Bridge] Add a DateInterval form type

### PR DESCRIPTION
**Pull request by MisatoTremor**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #13389 |
| License | MIT |
| Doc PR | symfony/symfony-docs#4817 |

### CODE REVIEW

**Comment by MisatoTremor** [comment]
Travis tests only failing temporarily (seems this is caused by #13351 and fixed soon).

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            $result[$field] = $dateInterval->format('%'.($this->pad ? strtoupper($char) : $char));
+        }
+        if (in_array('weeks', $this->fields, true)) {
+            $result['weeks'] = 0;
+            if (isset($result['days']) && (int) $result['days'] >= 7) {
+                $result['weeks'] = (string) floor($result['days'] / 7);
+                $result['days'] = (string) ($result['days'] % 7);
+            }
+        }
+        $result['invert'] = $result['invert'] === '-' ? true : false;
```
Comment (stof) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php]: just ``$result['invert'] = $result['invert'] === '-'`` will do it. ``$result['invert'] === '-'`` already gives you a boolean

Comment (MisatoTremor) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php]: Yep, thanks!

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    {
+        if (null === $fields) {
+            $fields = array('years', 'months', 'days', 'hours', 'minutes', 'seconds', 'invert');
+        }
+        $this->fields = $fields;
+        $this->pad = (bool) $pad;
+    }
+
+    /**
+     * Transforms a normalized date interval into a interval array.
```
Comment (hhamon) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php]: `a interval array` => `an interval array`

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+     *
+     * @param  \DateInterval                 $dateInterval Normalized date interval.
+     * @return array                         Interval array.
+     * @throws TransformationFailedException If the given value is not an
+     *                                                    instance of \DateInterval.
```
Comment (hhamon) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php]: I guess you can keep it on the same line.

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+            $result[$field] = $dateInterval->format('%'.($this->pad ? strtoupper($char) : $char));
+        }
+        if (in_array('weeks', $this->fields, true)) {
+            $result['weeks'] = 0;
+            if (isset($result['days']) && (int) $result['days'] >= 7) {
+                $result['weeks'] = (string) floor($result['days'] / 7);
+                $result['days'] = (string) ($result['days'] % 7);
+            }
+        }
+        $result['invert'] = $result['invert'] === '-';
```
Comment (hhamon) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php]: `$result['invert'] = '-' === $result['invert'];`

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToStringTransformer.php [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+/**
+ * Transforms between a date string and a DateInterval object
+ *
+ * @author Steffen Roßkamp <steffen.rosskamp@gimmickmedia.de>
+ */
+class DateIntervalToStringTransformer implements DataTransformerInterface
+{
+    /**
+     * Format used for generating strings
+     * @var string
```
Comment (hhamon) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToStringTransformer.php]: Missing blank line.

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToStringTransformer.php [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            throw $e;
+        } catch (\Exception $e) {
+            throw new TransformationFailedException($e->getMessage(), $e->getCode(), $e);
+        }
+
+        return $dateInterval;
+    }
+
+    /**
+     * @param $string
```
Comment (hhamon) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToStringTransformer.php]: PHPDoc can be removed in my opinion.

**Comment by MisatoTremor** [comment]
Updated deprecated functions and rebased.

**Comment by MisatoTremor** [comment]
@stof As 2.7 is now in feature freeze, should I rebase this to 3.0?

**Comment by WouterJ** [comment]
@MisatoTremor you should rebase on 2.8 now

**Comment by MisatoTremor** [comment]
Done
@WouterJ Is it correct this way or should I open a new PR based on 2.8 (and with squashed commits) or else?

**Comment by webmozart** [comment]
Thanks for creating this PR @MisatoTremor! Yes, you should open a new PR based on 2.8, otherwise it's hard to review this. :)

**Comment by MisatoTremor** [comment]
@webmozart ^ Done :)

**Pull request closed without merging**